### PR TITLE
Volttron secure mode: Run agent as separate user and restrict permissions to agents. 

### PR DESCRIPTION
**Pull request by schandrika**
New feature. Agents now run as separate user and have restricted read/write access.  In this mode, agents can write only to <install-dir>/<agent-name>.agent-data folder. 

Tested using new test module -  test_aip_security.py

Created new documentation about secure mode.

### CODE REVIEW

**Comment by craig8** [comment]
Working through the documentation on creating a secure volttron.  There is no mention of "creating a user" just that we are prompted to figure out what user to use.  That seems like a missing step to me.
image

**Comment by craig8** [comment]
Allow variable expansion...since I set my VOLTTRON_HOME to be the secure environment already we should be able to use it 
image

Also add a space and/or don't have so many words on lines...very claustrophobic!

**Comment by craig8** [comment]
After running the script to enable the secure mode the following error occured when running vcfg 
image

**Comment by craig8** [comment]
Merge with latest develop and hopefully some of the debug level logging pictured here will go away
image

**Comment by craig8** [comment]
Curious as to why we are missing a HEALTH for master driver.
image

**Comment by craig8** [comment]
I think this is very close to being ready...There are some goofy formatting issues in the docs and the missing "create this user" kinda stuff but other than that this is really good.

**Comment by schandrika** [comment]
1. Platform user should be an existing user. I will update the documentation for this
2. Will add code to expand variables
3. This prompt shouldn't be even there in vcfg any more. It is now part of the secure_user_permissions.sh script. I will fix this. 
4 & 5 - Will check on these. 
Thanks for reviewing this Craig.

**Comment by craig8** [comment]
@schandrika let me know and I will give it another go.

**Comment by schandrika** [comment]
> Allow variable expansion...since I set my VOLTTRON_HOME to be the secure environment already we should be able to use it
> image
> 
> Also add a space and/or don't have so many words on lines...very claustrophobic!

On second thought I think we should allow variable substitutions since this script will be run as sudo or root and hence the env variables will be different. This could cause confusion.

**Comment by schandrika** [comment]
@craig8 @shwethanidd I have updated code and documentation based on review comments and testing on a new environment. Please give it a try and let me if this looks ready to merge

File: docs/source/core_services/security/running_agent_as_user.rst [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+limited sudo access to create and delete agent users.
+
+Since this feature require system level changes (sudo access, user creation, file permission changes) the initial step
+needs to be run as root or user with sudo access. This can be a user other than Unix user used to run volttron platform.
+All files and folder created by VOLTTRON process in this mode would by default not have any access to others.
+Permission for Unix group others would be provided to specific files and folder based on VOLTTRON process requirement.
+It is recommended that you use a new volttron home to run volttron in secure mode. Converting a existing VOLTTRON
+instance to secure mode is also possible but would involve some manual changes. Please see section
+`Porting existing volttron home to secure mode`_.
+
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply docs/source/core_services/security/running_agent_as_user.rst]: done

File: docs/source/core_services/security/running_agent_as_user.rst [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+-------------------
+
+In this secure mode, agents will only have read write access to agent-data directory under the agent install
+directory - VOLTTRON_HOME/agents/<agent_uuid>/<agent_name>/<agent_name>.agent-data. Attempting to write in any other
+folder under VOLTTRON_HOME will result in permission errors.
+
+Changes to existing agents in secure mode
+-----------------------------------------
+
+Due to the above change, **SQL historian has been modified to create its database by default under its agent-data directory**
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply docs/source/core_services/security/running_agent_as_user.rst]: done

**Pull request merged**